src/transform/tests/testdata/join-implementation

```diff
@@ -34,22 +34,22 @@ opt
 (join [(get x)] [[#0 #1 #2]])
 ----
 Project (#0..=#2)
-  Filter ((#5 AND #7 AND (#6 OR (#3 AND #4))) OR ((#0) IS NULL AND ((#3 AND (#4 OR (#5 AND #6))) OR (#4 AND #6 AND #7))))
-    Map ((#1) IS NULL, (#2) IS NULL, (#0 = #2), (#1 = #2), (#0 = #1))
+  Filter ((#4 AND #5) OR ((#0) IS NULL AND ((#3 AND #5) OR ((#1) IS NULL AND (#3 OR #4)))))
+    Map ((#2) IS NULL, (#0 = #2), (#0 = #1))
       Get x
 
 opt
 (join [(get x)] [[#0 #2 #1 #2]])
 ----
 Project (#0..=#2)
-  Filter ((#5 AND #7 AND (#6 OR (#3 AND #4))) OR ((#0) IS NULL AND ((#3 AND (#4 OR (#5 AND #6))) OR (#4 AND #6 AND #7))))
-    Map ((#1) IS NULL, (#2) IS NULL, (#0 = #2), (#1 = #2), (#0 = #1))
+  Filter ((#4 AND #5) OR ((#0) IS NULL AND ((#3 AND #5) OR ((#1) IS NULL AND (#3 OR #4)))))
+    Map ((#2) IS NULL, (#0 = #2), (#0 = #1))
       Get x
 
 opt
 (join [(get x)] [[#0 #1 5]])
 ----
-Filter (#0 = 5) AND (#1 = 5)
+Filter (#0 = 5) AND (#0 = #1)
```

Comment:
Isn't this a regressions over the previous state?

Reply:
Yes, unfortunately this can conceivably introduce a regression. With the original version, if we have an index on `x(#0, #1)`, then we can look up `(5,5)`, but the new filter can do a lookup only with an index `x(#0)`. However, this is a single-input join, which is kind of a niche situation, so I'm not too worried about it. This is somewhat evidenced by the fact that this didn't happen in any of our `EXPLAIN` tests with real queries, but only in this unit test, which starts not from a real query but from an unusual MIR plan.

In the medium to long term, we should prevent brittleness like this by more principled canonicalizing of equivalences. Here, we should canonicalize this set of predicates by relying on the transitivity of `=`, and preferring 5 over `#1`. `EquivalencePropagation` has some machinery for this, and I'm not really sure why it's not catching this already.

Reply:
I could maybe introduce a feature flag for this before the release cut, if we are too worried about this.

Reply:
Or, I'll try to fix this regression on Wednesday by making `unpack_equivalences` equate not always to the first expression in the equivalence class, but find the one that's the simplest in the ordering of `EquivalenceClasses::tidy` (and has a smaller index than the one we are looking at, to prevent the class from falling apart into two components).